### DISCUSSION TITLE
strange heading measurement

### DISCUSSION THREAD

**Discussion by michelvancan** [pybricks/support#1811] [General]
I am using Audi e-tron Lego 42160 with technic hub and imu.heading() function with the following program
which makes the car do circles: 

```
def circle(speed):
    car.steer(-100)
    wait(1000)
    horloge = StopWatch()
    car.drive_power(speed)
    while True:
        temps = horloge.time()
        cap = -hub.imu.heading()
        print(f"{temps}, cap = {cap:.1f}, {hub.imu.angular_velocity()[2]/20:.1f}")
        if cap > 720: break
        wait(50)
    front.hold()
    rear.hold()

circle(50)
```

1) with "car.drive_power(50)" I get wrong heading measurements and the slower the speed the more wrong it is!
2) with "car.drive_speed(800)" the heading measurements are correct

Why?
I was using "car.drive_power" for slow speeds, because "car.drive_speed" works badly with low speeds (hiccups)
I encountered this problem when programming the odometry (which works relatively well).

Thanks for your help

**Comment by laurensvalk** [comment]
What do you mean that the heading is wrong? In which way is it different between the methods?

**Reply by michelvancan** [reply]
When the car stops he has made 2 1/4 turns. The error is 90 degrees.

**Comment by michelvancan** [comment]
I think I solved my problem, it doesn't depend on `car.drive_power` or `car.drive_speed`. Actually 
`print(hub.imu.settings())` gives `(4.97, 2498.96) `which is too high. So I added:
`hub.imu.settings(0, 0) `and it now works correctly!! (an IMU initialization problem?)

Thank you for this Pybricks software which opens up multiple perspectives to LEGO models. 
I will now be able to test my odometry program.

**Reply by laurensvalk** [reply]
The gyro needs a few seconds of being stationary at least once when you boot it up, to calibrate itself.

You can check `hub.imu.ready()` to see if this has happened yet.

**Reply by michelvancan** [reply]
No 
```
print(version)
#hub.imu.settings(0, 0)
wait(2000)
print(hub.imu.settings(), hub.imu.ready())
```
gives 
```
('technichub', '3.5.0', 'v1.20.0-23-g6c633a8dd on 2024-04-11')
(4.97, 2498.96) True
```

**Reply by laurensvalk** [reply]
Right. And did it exhibit the problem after it said that?

**Reply by michelvancan** [reply]
Yes, `hub.imu.settings(0, 0)` solve my problem. 
But in the manual you can read:
```
Parameters
 :
angular_velocity_threshold ([Number](https://code.pybricks.com/static/docs/v2.19.0/signaltypes.html#numbers), deg/s) – 
The threshold for angular velocity. The default value is 1.5 deg/s.

acceleration_threshold ([Number](https://code.pybricks.com/static/docs/v2.19.0/signaltypes.html#numbers), mm/s²) – 
The threshold for angular velocity. The default value is 250 mm/s².
```
which is not the case

**Reply by laurensvalk** [reply]
> `hub.imu.settings(0, 0)` solve my problem.

Are you definitely sure it was that, or could the hub just have been sitting still for a few seconds as you were making that change?

Those settings configure below which values the sensor considers itself steady enough to calibrate. If you set them to 0, it isn't going to calibrate at all.

**Reply by michelvancan** [reply]
I am sure!!!. Effectively it gives False for `hub.imu.stationary(),` but the **heading is correct** after 2 full turns. 
I have changed the setting to hub.imu.settings(1, 100) and so stationery go to True and it give also the correct heading

here is the complete program
```
from pybricks.hubs import TechnicHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import Car
from pybricks.tools import wait, StopWatch


hub = TechnicHub()
hub.imu.settings(0, 0)
wait(1000)
print(hub.imu.settings(), hub.imu.ready(), hub.imu.stationary())
hub.imu.reset_heading(0)
wait(3000)
print(hub.imu.settings(), hub.imu.ready(), hub.imu.stationary())

# Set up all devices.
steering = Motor(Port.D, Direction.CLOCKWISE)
front = Motor(Port.B, Direction.CLOCKWISE)
rear = Motor(Port.A, Direction.CLOCKWISE)
car = Car(steering, [front, rear])
wait(1000)

def circle(speed):
    car.steer(-100)
    wait(1000)
    horloge = StopWatch()
    car.drive_power(speed)
    while True:
        temps = horloge.time()
        cap = -hub.imu.heading()
        print(f"{temps}, cap = {cap:.1f}, {hub.imu.angular_velocity()[2]/20:.1f}")
        if cap > 720: break
        wait(50)
    front.hold()
    rear.hold()

circle(50)
```
and the results 
```
(0.07, 2.39364) True False
(0.07, 2.39364) True False
0, cap = 0.1, 0.0
50, cap = 0.1, 0.0
100, cap = 0.2, 0.1
150, cap = 0.5, 0.6
200, cap = 1.3, 1.1
250, cap = 2.5, 1.3
300, cap = 3.9, 1.5
350, cap = 5.3, 1.4
400, cap = 6.7, 1.4
450, cap = 8.2, 1.5
......
24667, cap = 716.6, 1.5
24717, cap = 718.0, 1.4
24767, cap = 719.4, 1.4
24817, cap = 720.8, 1.4
```

**Reply by laurensvalk** [reply]
Thank you for the details. I'm glad you have found something that works. I was just hoping to reproduce the case that _doesn't work_, so we can try to identify and fix it for others :slightly_smiling_face: 

So for example, could you share the script where it does produce the error?

Thanks!

**Reply by michelvancan** [reply]
I think the error appears with this little program 
```
from pybricks.tools import wait, StopWatch
from pybricks.hubs import ThisHub 
from pybricks import version

# Initialize the hub.
hub = ThisHub()
print(version)
wait(1000)
print(hub.imu.settings(), hub.imu.ready())

('technichub', '3.5.0', 'v1.20.0-23-g6c633a8dd on 2024-04-11')
(4.97, 2498.96) True
```
The default values ​​are too large and the gyroscope misses rotations because it thinks it is standing still when it is not.
You absolutely must use  : for `example hub.imu.settings(1, 100)`

**Comment by BertLindeman** [comment]
Laurens,

Do not know if this helps, but I did some measurements with various hubs / versions.
The result of CI build 3540 stands out. Very low settings and the imu never (waited over a minute) reaches ready or stationary.
There is no difference if I do ```hub = ThisHub()``` or ```hub = PrimeHub()``` or another type of hub.

The result:

```
define ThisHub 
('primehub', '3.5.0', 'ci-build-3540-v3.5.0-90-g9d3a6f85 on 2024-09-06')
(0.07, 2.39364, 361.3) ready:False stationary:False

define PrimeHub 
('primehub', '3.5.0', 'ci-build-3540-v3.5.0-90-g9d3a6f85 on 2024-09-06')
(0.07, 2.39364, 361.3) ready:False stationary:False

define TechnicHub 
('technichub', '3.5.0', 'v1.20.0-23-g6c633a8dd on 2024-04-11')
(4.97, 2498.96) ready:True stationary:True

define TechnicHub 
('technichub', '3.5.0', 'ci-build-3542-v3.5.0-91-g068d2cae on 2024-09-07')
(4.97, 2498.96, 364.5359) ready:True stationary:True

define PrimeHub 
('primehub', '3.5.0', 'ci-build-3542-v3.5.0-91-g068d2cae on 2024-09-07')
(4.97, 2498.96, 360.0) ready:True stationary:True
```

The program printed for a few seconds, but as the results are always the same, I removed the extra print output.

<details><summary>Test program and snippet for build 3540 </summary>

```python

from pybricks.tools import wait, StopWatch
from pybricks.hubs import PrimeHub 
from pybricks import version

# Initialize the hub.
print(f'define PrimeHub ')
hub = PrimeHub()
print(version)
wait(1000)
print(f'{hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')
wait(1000)
print(f'{hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')
wait(1000)
print(f'{hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')
wait(1000)
print(f'{hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')
wait(1000)
print(f'{hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')
wait(1000)
print(f'{hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')
```

```python
# Snippet to wait for the imu to get ready (it does NOT)

while not hub.imu.ready():
    wait(1000)
    print(f'{watch.time() % 1000} sec {hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')

print(f'{watch.time() % 1000} sec {hub.imu.settings()} ready:{hub.imu.ready()} stationary:{hub.imu.stationary()}')
```
</details>

Bert

**Reply by laurensvalk** [reply]
Thanks @BertLindeman . I think OP is using the stable firmware. There has been some ongoing work on the IMU since then, so I think that is unrelated to this question. Based on what I have seen, everything did seem to work after the hub was sitting still for a little while.

(The change you are seeing is that the settings will now be persistent across reboots, which should help with this question / issue. But it looks like there may still be some data on your hub, just as you saw with the program slots. I think we need to make sure to properly erase this for people using beta versions, not just clear it between firmware updated.)
